<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Bionformatics, Matrix Multiplication and Hash Tables | Research Log</title>
<meta name="generator" content="Jekyll v4.3.2" />
<meta property="og:title" content="Bionformatics, Matrix Multiplication and Hash Tables" />
<meta name="author" content="SantosJGND" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="A common problem in bioinformatics is to count the number of times a sequence appears in a genome. For example, you might want to know how many times the sequence ATG appears in the genome of a given organism. This is a simple problem, but it can be computationally expensive. The genome of a bacteria can be millions of base pairs long, and the sequence you are looking for can be of any length. So how do you solve this problem efficiently?" />
<meta property="og:description" content="A common problem in bioinformatics is to count the number of times a sequence appears in a genome. For example, you might want to know how many times the sequence ATG appears in the genome of a given organism. This is a simple problem, but it can be computationally expensive. The genome of a bacteria can be millions of base pairs long, and the sequence you are looking for can be of any length. So how do you solve this problem efficiently?" />
<link rel="canonical" href="http://localhost:4000/bioinformatics/matrix-multiplication/hash-tables/counting/2023/07/19/bioinformatics-counts.html" />
<meta property="og:url" content="http://localhost:4000/bioinformatics/matrix-multiplication/hash-tables/counting/2023/07/19/bioinformatics-counts.html" />
<meta property="og:site_name" content="Research Log" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2023-07-19T00:34:26+01:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Bionformatics, Matrix Multiplication and Hash Tables" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"SantosJGND"},"dateModified":"2023-07-19T00:34:26+01:00","datePublished":"2023-07-19T00:34:26+01:00","description":"A common problem in bioinformatics is to count the number of times a sequence appears in a genome. For example, you might want to know how many times the sequence ATG appears in the genome of a given organism. This is a simple problem, but it can be computationally expensive. The genome of a bacteria can be millions of base pairs long, and the sequence you are looking for can be of any length. So how do you solve this problem efficiently?","headline":"Bionformatics, Matrix Multiplication and Hash Tables","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/bioinformatics/matrix-multiplication/hash-tables/counting/2023/07/19/bioinformatics-counts.html"},"url":"http://localhost:4000/bioinformatics/matrix-multiplication/hash-tables/counting/2023/07/19/bioinformatics-counts.html"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/css/style.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="Research Log" />


  
</head>
<body><header class="site-header">

  <div class="wrapper"><a class="site-title" rel="author" href="/">Research Log</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/about/">About</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Bionformatics, Matrix Multiplication and Hash Tables</h1>
    <p class="post-meta"><time class="dt-published" datetime="2023-07-19T00:34:26+01:00" itemprop="datePublished">
        Jul 19, 2023
      </time></p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <p>A common problem in bioinformatics is to count the number of times a sequence appears in a genome. For example, you might want to know how many times the sequence <code class="language-plaintext highlighter-rouge">ATG</code> appears in the genome of a given organism. This is a simple problem, but it can be computationally expensive. The genome of a bacteria can be millions of base pairs long, and the sequence you are looking for can be of any length. So how do you solve this problem efficiently?</p>

<h3 id="constructing-profiles">Constructing profiles</h3>

<p>Lets begin by how we record profiles - think of a record of the presence / absence of set of patterns in a given sequence. For example, for the sorted set <code class="language-plaintext highlighter-rouge">[A, T, C, G]</code>, we can represent the letter pattern <code class="language-plaintext highlighter-rouge">A</code> as <code class="language-plaintext highlighter-rouge">[1, 0, 0, 0]</code>, the letter pattern <code class="language-plaintext highlighter-rouge">T</code> as <code class="language-plaintext highlighter-rouge">[0, 1, 0, 0]</code>, and so on. We can then represent the sequence <code class="language-plaintext highlighter-rouge">ATG</code> as <code class="language-plaintext highlighter-rouge">[[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1]]</code>.</p>

<p>You may be tempted to use generate these profiles using membership tests, as follows:</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">pattern</span><span class="o">=</span> <span class="p">[</span><span class="s">"A"</span><span class="p">]</span>
<span class="n">sorted_set</span> <span class="o">=</span> <span class="p">[</span><span class="s">"A"</span><span class="p">,</span> <span class="s">"T"</span><span class="p">,</span> <span class="s">"C"</span><span class="p">,</span> <span class="s">"G"</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">count_patterns</span><span class="p">(</span><span class="n">pattern</span><span class="p">,</span> <span class="n">sorted_set</span><span class="p">):</span>
    
    <span class="k">return</span> <span class="p">[</span><span class="mi">1</span> <span class="k">if</span> <span class="n">letter</span> <span class="ow">in</span> <span class="n">pattern</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">letter</span> <span class="ow">in</span> <span class="n">sorted_set</span><span class="p">]</span>
    </code></pre></figure>

<p>But this is not efficient. The <code class="language-plaintext highlighter-rouge">in</code> operator will loop through the entire sequence, and for every element in the sequence it will loop through the entire pattern.</p>

<p>Alternatively, we can use hash-tables (dictionaries in python):</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">pattern</span><span class="o">=</span> <span class="p">{</span><span class="s">"A"</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span>
<span class="n">sorted_set</span> <span class="o">=</span> <span class="p">[</span><span class="s">"A"</span><span class="p">,</span> <span class="s">"T"</span><span class="p">,</span> <span class="s">"C"</span><span class="p">,</span> <span class="s">"G"</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">count_patterns</span><span class="p">(</span><span class="n">pattern</span><span class="p">,</span> <span class="n">sorted_set</span><span class="p">):</span>
    
    <span class="k">return</span> <span class="p">[</span>
        <span class="n">pattern</span><span class="p">.</span><span class="nf">get</span><span class="p">(</span><span class="n">letter</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">letter</span> <span class="ow">in</span> <span class="n">sorted_set</span>
    <span class="p">]</span>
    </code></pre></figure>

<p>This is more efficient. The <code class="language-plaintext highlighter-rouge">get</code> method of dictionaries has a time complexity of <code class="language-plaintext highlighter-rouge">O(1)</code>, so this will be faster than the membership test. The difference is not noticeable for small sequences, but it becomes noticeable for large sequences. Counting occurrences could be with little modifications.</p>

<h3 id="counting-shared-occurrences">Counting shared occurrences</h3>

<p>The naive approach is to loop through the sequence, and for every position in the sequence, check if the pattern appears in another. This is not efficient. For a sequence of length <code class="language-plaintext highlighter-rouge">n</code>, and a pattern of length <code class="language-plaintext highlighter-rouge">k</code>, this has a time complexity of <code class="language-plaintext highlighter-rouge">O(nk)</code>.</p>

<p>Lets try to solve this problem using matrix multiplication. Letâ€™s say we have constructed a list of profiles,</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">profiles</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="p">]</span></code></pre></figure>

<p>And we wish to create a matrix of shared occurrences. We can do this by multiplying the transpose of the profiles with the profiles themselves, with the resulting product:</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">product</span><span class="o">=</span> <span class="p">[</span>
    <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
<span class="p">]</span></code></pre></figure>

<p>where the diagonal will represent the sum of patterns in each profile, and the off-diagonal elements will represent the number of shared occurrences between profiles.</p>

<p>this is easy in python using the numpy package:</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">profiles</span><span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="nf">array</span><span class="p">([</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="p">])</span>


<span class="n">product</span> <span class="o">=</span> <span class="n">profiles</span> <span class="o">@</span> <span class="n">profiles</span><span class="p">.</span><span class="n">T</span></code></pre></figure>

<p>This is much more efficient. The time complexity of matrix multiplication is <code class="language-plaintext highlighter-rouge">O(n^3)</code>, so this is much faster than the naive approach.</p>

<h3 id="very-large-matrices">very large matrices</h3>

<p>While matrix multiplication is efficient, it is not practical to keep entire matrices in memory. The solution is to split these operations into chunks, e.g.:</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="k">def</span> <span class="nf">chunk_matrix</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">chunk_size</span><span class="p">):</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nf">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">matrix</span><span class="p">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">chunk_size</span><span class="p">):</span>
        <span class="k">yield</span> <span class="n">matrix</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">chunk_size</span><span class="p">,</span> <span class="p">:]</span>


<span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="nf">chunk_matrix</span><span class="p">(</span><span class="n">profiles</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>

    <span class="n">subset_product</span><span class="o">=</span> <span class="n">chunk</span> <span class="o">@</span> <span class="n">profiles</span><span class="p">.</span><span class="n">T</span>

    <span class="c1">## do something with subset_product</span></code></pre></figure>

<h1 id="conclusion">Conclusion</h1>

<p>This is a simple idea with many applications. For example i have used it to extract the mutation spectrum in primates [1], more recently to study patterns of cross-mapping in metagenomics [2]. It occurs me that it could be useful to apply to VCF files given their binary nature.</p>

<p>[1] https://doi.org/10.1093/gbe/evad019</p>

<p>[2] https://insaflu.insa.pt/</p>


  </div><a class="u-url" href="/bioinformatics/matrix-multiplication/hash-tables/counting/2023/07/19/bioinformatics-counts.html" hidden></a>
</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <div class="footer-col-wrapper">
      <div class="footer-col">
        <ul class="contact-list">
          <li class="p-name">SantosJGND</li>
          <li>
                <i class="fa fa-github" aria-hidden="true"></i>
                <a class="u-url" href="https://github.com/SantosJGND">SantosJGND</a>
            </li>
            <li>
                <i class="fa fa-twitter" aria-hidden="true"></i>
                <a class="u-url" href="https://twitter.com/SantosJGND">SantosJGND</a>
            </li>
            <li>
                <i class="fa fa-envelope" aria-hidden="true"></i>
                <a class="u-email" href="mailto:dourado.jns@gmail.com">dourado.jns@gmail.com</a>
            </li>
        </ul>
      </div>
      <div class="footer-col">
        <p>A technical account of my research path where I expose the problems I faced and how I solved them, with some detours into interesting tangents along the way.
</p>
      </div>
    </div>

    <div class="social-links"><ul class="social-media-list"></ul>
</div>

  </div>

</footer>
</body>

</html>

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
<link rel="shortcut icon" type="image/png" href='/images/dna.png'>